### PR TITLE
feat: support icicle api for all the fields

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This enables ICICLE APIs for all the fields.

### CODE REVIEW

**Review by GideokKim — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by batzor — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: tachyon/math/polynomials/univariate/icicle/icicle_ntt_bn254.cc [hunk 1]
```diff
@@ -18,3 +24,95 @@ cudaError_t tachyon_bn254_ntt_cuda(const ::bn254::scalar_t* input, int size,
...
+template <>
+bool IcicleNTT<bn254::Fr>::Init(const bn254::Fr& group_gen,
+                                const IcicleNTTOptions& options) {
+#if FIELD_ID != BN254
+#error Only BN254 is supported
+#endif
+  ::device_context::DeviceContext ctx{stream_, /*device_id=*/0, mem_pool_};
+  math::BigInt<4> group_gen_big_int = group_gen.ToBigInt();
+  // TODO(chokobole): There are some issues about domain initialization, we need
+  // to handle each of these.
```
Comment (ashjeong) [inline tachyon/math/polynomials/univariate/icicle/icicle_ntt_bn254.cc:39]: Please change in all related files
```suggestion
  // TODO(chokobole): We must handle these issues with domain initialization:

File: tachyon/math/polynomials/univariate/icicle/icicle_ntt_bn254.cc [hunk 2]
```diff
@@ -18,3 +24,95 @@ cudaError_t tachyon_bn254_ntt_cuda(const ::bn254::scalar_t* input, int size,
...
+  // 1. It gets too slow when the domain size is small, 1, 2, and 4.
```
Comment (ashjeong) [inline tachyon/math/polynomials/univariate/icicle/icicle_ntt_bn254.cc:40]:
```suggestion
  // 1. It gets too slow when the domain size is 1, 2, 4, or small in general.

File: tachyon/math/polynomials/univariate/icicle/icicle_ntt_bn254.cc [hunk 3]
```diff
@@ -18,3 +24,95 @@ cudaError_t tachyon_bn254_ntt_cuda(const ::bn254::scalar_t* input, int size,
...
+  //    See "vendors/circom/prover_main.cc".
+  // 2. |fast_twiddles_mode| consumes a lot of memory, so we need to disable if
```
Comment (ashjeong) [inline tachyon/math/polynomials/univariate/icicle/icicle_ntt_bn254.cc:42]:
```suggestion
  // 2. |fast_twiddles_mode| consumes a lot of memory, so we need to disable it if

File: tachyon/math/polynomials/univariate/icicle/icicle_ntt_bn254.cc [hunk 4]
```diff
@@ -18,3 +24,95 @@ cudaError_t tachyon_bn254_ntt_cuda(const ::bn254::scalar_t* input, int size,
...
+bool IcicleNTT<bn254::Fr>::Run(::ntt::NttAlgorithm algorithm,
+                               const BigInt& coset, bn254::Fr* inout, int size,
+                               ::ntt::NTTDir dir) const {
+#if FIELD_ID != BN254
+#error Only BN254 is supported
+#endif
+
+  // NOTE(chokobole): Manual copy is needed even though
+  // |sizeof(::bn254::scalar_t)| and |sizeof(bn254::Fr)| are same. This is
+  // because their alignment are different.
```
Comment (ashjeong) [inline tachyon/math/polynomials/univariate/icicle/icicle_ntt_bn254.cc:78]:
```suggestion
  // because their alignments are different.

**Review by ashjeong — APPROVED** [review]
LGTM

**Pull request merged**